common/src/main/java/io/pravega/common/util/SortedDeque.java

```diff
@@ -0,0 +1,269 @@
+        SortedIndex.IndexEntry last = this.elements[(this.tail - 1) & this.lastSlotIndex];
+        Preconditions.checkArgument(isEmpty() || item.key() > last.key(), "keys must be in increasing order.");
+        this.elements[this.tail] = item;
+        this.tail = (this.tail + 1) & this.lastSlotIndex;
+        if (this.tail == this.head) {
+            doubleCapacity();
+        }
+    }
+
+    /**
+     * Removes the first item in the Deque.
+     *
+     * @return The first item in the Deque, or null if Deque is empty.
+     */
+    public E removeFirst() {
+        E result = (E) this.elements[this.head]; // Element is null if deque empty
+        if (result == null) {
+            return null;
+        }
+
+        // Drop the pointer to the result and move the head over by one slot.
+        this.elements[this.head] = null;
+        this.head = (this.head + 1) & this.lastSlotIndex;
+        return result;
+    }
+
+    /**
+     * Removes all the items from the beginning of the Deque up to and including the given item, but ONLY if the given
+     * item exists in the Deque.
+     *
+     * @param upToIncluding The Item to remove up to and including. If this item does not exist, no change is made.
+     * @return True if any changes were made (and thus upToIncluding exists), false otherwise.
+     */
+    public boolean removeFirst(E upToIncluding) {
+        if ((this.head != this.tail) && upToIncluding.key() == this.elements[this.head].key()) {
+            // Quick check: our desired item is the first one; don't bother doing a binary search.
+            removeFirst();
+            return true;
+        }
+
+        int itemIndex = binarySearch(upToIncluding.key());
+        if (itemIndex < 0) {
+            //Item does not exist, as such we cannot remove anything.
+            return false;
+        }
+
+        // Remove all items from the beginning, up to and including our sought item.
+        while (itemIndex-- >= 0) {
+            this.elements[this.head] = null;
+            this.head = (this.head + 1) & this.lastSlotIndex;
+        }
+
+        return true;
+    }
+
+    /**
+     * Removes the last item in the Deque.
+     *
+     * @return The last item in the Deque, or null if Deque is empty.
+     */
+    public E removeLast() {
+        int t = (this.tail - 1) & this.lastSlotIndex;
+        E result = (E) this.elements[t];
+        if (result == null) {
+            return null;
+        }
+
+        this.elements[t] = null;
+        this.tail = t;
+        return result;
+    }
+
+    /**
+     * Removes all the items from the end of the Deque up to and including the given item, but ONLY if the given item
+     * exists in the Deque.
+     *
+     * @param fromIncluding The Item to remove up to and including. If this item does not exist, no change is made.
+     * @return True if any changes were made (and thus fromIncluding exists), false otherwise.
+     */
+    public boolean removeLast(E fromIncluding) {
```

Comment:
1. Renamed.
2. Yes, this is the exact behavior I want. The reason is that this is used to store checkpoints. **Before** I make any changes to the collection, I want to make sure my checkpoint exists. Otherwise I could unintentionally modify it in case of an error and lose data. If I didn't need this constraint then I wouldn't need this whole method, let alone the binary search part of it.